text/0000-time-improvements.md

```diff
@@ -0,0 +1,327 @@
+
+```rs
+impl SystemTime {
+  /// Panics if `earlier` is later than &self.
+  /// Because SystemTime is monotonic, the only ime that `earlier` should be
+  /// a later time is a bug in your code.
+  pub fn duration_from_earlier(&self, earlier: SystemTime) -> SystemTime;
+
+  /// Panics if self is later than the current time (can happen if a SystemTime
+  /// is produced synthetically)
+  pub fn elapsed(&self) -> Duration;
+}
+
+impl Add<Duration> for SystemTime {
+  type Output = SystemTime;
+}
+
+impl Sub<Duration> for SystemTime {
+  type Output = SystemTime;
+}
+
+impl PartialEq for SystemTime;
+impl Eq for SystemTime;
+impl PartialOrd for SystemTime;
+impl Ord for SystemTime;
+```
+
+For convenience, several new constructors are added to `Duration`. Because any
+unit greater than seconds has caveats related to leap seconds, all of the
+constructors take "standard" units. For example a "standard minute" is 60
+seconds, while a "standard hour" is 3600 seconds.
+
+The "standard" terminology comes from [JodaTime][joda-time-standard].
+
+[joda-time-standard]: http://joda-time.sourceforge.net/apidocs/org/joda/time/Duration.html#standardDays(long)
+
+```rs
+impl Duration {
+  /// a standard minute is 60 seconds
+  /// panics if the number of minutes is larger than u64 seconds
+  pub fn from_standard_minutes(minutes: u64) -> Duration;
+
+  /// a standard hour is 60 standard minutes
+  /// panics if the number of hours is larger than u64 seconds
+  pub fn from_standard_hours(hours: u64) -> Duration;
+
+  /// a standard day is 24 standard hours
+  /// panics if the number of days is larger than u64 seconds
+  pub fn from_standard_days(days: u64) -> Duration;
+}
+```
+
+### SystemTime
+
+**This type should not be used for in-process timestamps, like those used in
+benchmarks.**
+
+A `SystemTime` represents a time stored on the local machine derived from the
+system clock. For example, it is used to represent `mtime` on the file system.
+
+The most important caveat of `SystemTime` is that it is **not monotonic**. This
+means that you can save a file to the file system, then save another file to
+the file system, **and the second file has an `mtime` earlier than the second**.
+
+> **This means that an operation that happens after another operation in real
+> time may have an earlier `SystemTime`!**
+
+In practice, most programmers do not think about this kind of "time travel"
+with the system clock, leading to strange bugs once the mistaken assumption
+propagates through the system.
+
+This design attempts to help the programmer catch the most egregious of these
+kinds of mistakes (unexpected travel **back in time**) before the mistake
+propagates.
+
+```rs
+impl SystemTime {
+  /// Returns an `Err` if `earlier` is later
+  pub fn duration_from_earlier(&self, earlier: SystemTime) -> Result<Duration, ()>;
+
+  /// Returns an `Err` if &self is later than the current system time.
+  pub fn elapsed(&self) -> Result<Duration, ()>;
```

Comment:
`Err = ()` is not sufficient, because system time APIs might error-out for their own reasons.